Nil checks in iOS bridge for outcomes methods

rgomezp opened the pull request:
Motivation: there are multiple places in native `OneSignalOutcomeEventsController` where we call `success(nil)`. One example is calling `sendUniqueOutcome` multiple times. We should check that the `outcome` object is non-nil before passing it back to JS with the `jsonRepresentation` native method.

rgomezp committed:
Nil checks in iOS bridge for outcomes methods

Motivation: there are multiple places in native `OneSignalOutcomeEventsController` where we call `success(nil)`. One example is calling `sendUniqueOutcome` multiple times. We should check that the `outcome` object is non-nil before passing it back to JS with the `jsonRepresentation` native method.

Jeasmine commented on ios/RCTOneSignal/RCTOneSignalEventEmitter.m:
```diff
@@ -389,19 +389,25 @@ + (void)sendEventWithName:(NSString *)name withBody:(NSDictionary *)body {
 RCT_EXPORT_METHOD(sendOutcome:(NSString *)name :(RCTResponseSenderBlock)callback) {
     [OneSignal sendOutcome:name onSuccess:^(OSOutcomeEvent *outcome){
-        callback(@[[outcome jsonRepresentation]]);
+        if (outcome) {
```
nil outcome represents a failure, shouldn't we return a failure text o something?

rgomezp replied:
not currently. we are calling `success(nil)` from the native side so it is technically considered a success. The JS-side function currently takes one callback which expects an `OutcomeEvent` object as you can tell by the TS typings. 

There are two ways to handle callback errors in JS [as documented](https://reactnative.dev/docs/native-modules-ios#callbacks):

1. expect two arguments in the callback e.g: `(error, outcomeEvent)`
2. expect two separate callbacks e.g: `(outcomeEvent) => {}, (error) => {}`

out of the two options, I would prefer the second one. However, this would require a change to the typings and documentation. Also, from inspecting the [native code](https://github.com/OneSignal/OneSignal-iOS-SDK/blob/master/iOS_SDK/OneSignalSDK/Source/OneSignalOutcomeEventsController.m) it seems that the call to `success(nil)` really only happens with unique outcomes as far as I can tell. So that means that we would only really need to add error handling in the bridge for `sendUniqueOutcome`. But adding it only to that function would not be consistent.

For the above reasons, I suggest we hold off adding any error handling or any changes to the bindings themselves until we add a `onFailure` block native side (TO DO). 

As a sort of compromise we can add native-side logging for the nil case but this won't be visible to JS developers unless they open up Xcode.

jkasten2 replied:
I agree, there isn't a need to add to a signature for failures now if this isn't something both the OneSignal-Android-SDK and OneSignal-iOS-SDK can provide.

rgomezp committed:
Add native logging to outcomes methods in iOS bridge

Motivation: 
- [Context](https://bit.ly/2WJbxHG)
- Adds some logging in the iOS bridge to address the case that `outcome` object is nil

rgomezp commented on ios/RCTOneSignal/RCTOneSignalEventEmitter.m:
```diff
@@ -389,19 +389,34 @@ + (void)sendEventWithName:(NSString *)name withBody:(NSDictionary *)body {
+            return;
+        }
+
+        [OneSignal onesignalLog:ONE_S_LL_VERBOSE message:[NSString stringWithFormat:@"sendOutcome OSOutcomeEvent is nil."]];
```
Added logging to outcomes methods

jkasten2 reviewed (approved):
The iOS changes look good, however we need to update the Android logic to match.
https://github.com/OneSignal/react-native-onesignal/blob/f8a8e018cf1597bd38a2ee1cd0f9c885ffc3dade/android/src/main/java/com/geektime/rnonesignalandroid/RNOneSignal.java#L534-L550

Feel free to do this in this PR or a follow up PR.

rgomezp commented:
will open a new PR for Android
